store/filestore.go

```diff
@@ -39,11 +42,79 @@ func NewFileStore(root string, maxSize int64) *FileStore {
 	if _, err := os.Stat(root); os.IsNotExist(err) {
 		os.MkdirAll(root, 0755)
 	}
-	return &FileStore{
+	fs := &FileStore{
 		root:     root,
 		metadata: collections.NewSyncMap(),
 		maxSize:  maxSize,
 	}
+	fs.startSubDirectoriesFileWatcher()
+	return fs
+
+}
+
+func (fs *FileStore) startSubDirectoriesFileWatcher() {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// iterate over all subdirectories under root path
+	if err := filepath.Walk(fs.root,
+		func(path string, fi os.FileInfo, err error) error {
+			if fi.Mode().IsDir() {
+				return watcher.Add(path)
+			}
+			return nil
+		}); err != nil {
+
+		log.Fatalf("ERROR: %s", err)
+	}
+
+	go func() {
+		for {
+
+			// gather events
+			select {
+			case event := <-watcher.Events:
+				switch event.Op {
+
+				case fsnotify.Write:
+					buf, err := ioutil.ReadFile(event.Name)
+					if err != nil {
+						log.Fatalf(
+							"error while reading file: %s with error: %s",
+							event.Name, err)
+					}
+					size := int64(len(buf))
+
+					// update metadata
+					fs.metadata.Put(event.Name, file{filename: event.Name, size: size, atime: time.Now()})
+					atomic.AddInt64(&fs.size, size)
+
+				case fsnotify.Remove:
+					p := fs.metadata.Get(event.Name)
+					if p != nil {
+						m := p.(file)
+						atomic.AddInt64(&fs.size, -m.size)
+						fs.metadata.Remove(event.Name)
+					}
+
+				case fsnotify.Create:
+					fi, _ := os.Stat(event.Name)
+					if fi.Mode().IsDir() {
+						watcher.Add(event.Name)
```

Comment:
Given an empty thumbnail cache directory, after the following resize `/300x300/fit/0/example.jpg`, the watcher will only watch on the `300x300` subdir, whereas it should watch on:

- `300x300`
- `300x300/fit`
- `300x300/fit/0`

You should walk the subdirectory and call `watcher.Add` as many times as needed.

Reply:
It should also add `300x300/fit/0/example.jpg` to the metadata map.

Reply:
Seems done